lib/vnet/opensshconfig.go

```diff
@@ -0,0 +1,119 @@
+// third-party SSH clients can trust it. It then reads or generates
+// ${TELEPORT_HOME}/id_vnet(.pub) which SSH clients should be configured to use
+// for connections to VNet SSH. It returns id_vnet.pub so that VNet SSH can
+// trust it for incoming connections.
+func writeSSHKeys(homePath string, hostCAKey []byte) ([]byte, error) {
+	profilePath := profilePath(homePath)
+	if err := writeKnownHosts(profilePath, hostCAKey); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	userPubKey, err := readUserPubKey(profilePath)
+	if trace.IsNotFound(err) {
+		userPubKey, err = generateAndWriteUserKey(profilePath)
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return userPubKey, nil
+}
+
+func profilePath(homePath string) string {
+	if homePath == "" {
+		if homeDir := os.Getenv(types.HomeEnvVar); homeDir != "" {
+			homePath = filepath.Clean(homeDir)
+		}
+	}
+	return profile.FullProfilePath(homePath)
+}
+
+func writeKnownHosts(profilePath string, hostCAKey []byte) error {
+	p := keypaths.VNetKnownHostsPath(profilePath)
+	f, err := os.OpenFile(p, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, filePerms)
+	if err != nil {
+		return trace.Wrap(trace.ConvertSystemError(err), "opening %s for writing", p)
+	}
+	defer f.Close()
+	_, err = fmt.Fprintf(f, "@cert-authority * %s\n", strings.TrimSpace(string(hostCAKey)))
+	return trace.Wrap(trace.ConvertSystemError(err), "writing host CA to %s", p)
+}
+
+func readUserPubKey(profilePath string) ([]byte, error) {
+	p := keypaths.VNetClientSSHKeyPubPath(profilePath)
+	pubKey, err := os.ReadFile(p)
+	err = trace.ConvertSystemError(err)
+	return pubKey, trace.Wrap(err, "reading user public key from %s", p)
+}
+
+func generateAndWriteUserKey(profilePath string) ([]byte, error) {
+	userKey, err := cryptosuites.GenerateKeyWithAlgorithm(cryptosuites.Ed25519)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating SSH user key")
+	}
+
+	privPemBlock, err := ssh.MarshalPrivateKey(userKey, "")
+	if err != nil {
+		return nil, trace.Wrap(err, "marshaling SSH user key")
+	}
+	privKeyPath := keypaths.VNetClientSSHKeyPath(profilePath)
+	privKeyFile, err := os.OpenFile(privKeyPath, os.O_WRONLY|os.O_CREATE, filePerms)
+	if err != nil {
+		return nil, trace.Wrap(trace.ConvertSystemError(err), "opening %s for writing", privKeyPath)
+	}
+	defer privKeyFile.Close()
+	if err := pem.Encode(privKeyFile, privPemBlock); err != nil {
+		return nil, trace.Wrap(trace.ConvertSystemError(err), "writing user private key to %s", privKeyPath)
+	}
+
+	sshPub, err := ssh.NewPublicKey(userKey.Public())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	userPubKey := ssh.MarshalAuthorizedKey(sshPub)
+	pubKeyPath := keypaths.VNetClientSSHKeyPubPath(profilePath)
+	pubKeyFile, err := os.OpenFile(pubKeyPath, os.O_WRONLY|os.O_CREATE, filePerms)
+	if err != nil {
+		return nil, trace.Wrap(trace.ConvertSystemError(err), "opening %s for writing", pubKeyPath)
+	}
+	defer pubKeyFile.Close()
+	_, err = pubKeyFile.Write(userPubKey)
```

Comment:
Can we use `renameio.WriteFile` here and for `privKeyFile` so clients can't accidentally read partially written files? You're also not checking the error code of `Close()` so any IO problems are almost surely going to be ignored.

Reply:
Oh renameio doesn't work on Windows? https://github.com/google/renameio/issues/1